Feature/module gpiod

w4tsn opened the pull request:
This PR adds the gpiod (General Purpose Input Output Device) module (through libgpiod[1] and python3-libgpiod[2]) introduced with linux kernel 4.8.

The module uses the `/dev/gpiochipN` chardev to access the GPIOs of a system.

This module uses threading to add support for callback-style interrupts, which are not yet upstream.

Bias (Pullup) settings are introduced in kernel 5.5+ but are not yet part of the python3-libgpiod module. This renders it inferior to the sysfs based modules as of now.

Distributions like Fedora IoT ship without the deprecated sysfs interface `/sys/class/gpio...` and instead use the new kernel driver libgpiod.

The code was tested in it's current state with LEDs and Buttons on a Raspberry Pi 3 B+ with Fedora IoT 31.

Dependencies
---

The module requires libgpiod and libgpiod-devel (development files) on the host system.

It relies on python3-libgpiod.

References
---

[1] <https://git.kernel.org/pub/scm/libs/libgpiod/libgpiod.git>
[2] <https://pypi.org/project/gpiod/>
[3] <https://www.beyondlogic.org/an-introduction-to-chardev-gpio-and-libgpiod-on-the-raspberry-pi/>

w4tsn committed:
Add initial draft of gpiod module

w4tsn committed:
Rename the line consumer to 'pi-mqtt-gpio'

w4tsn committed:
Pass the cleanup

In theory everything should cleanup itself on exit. This is a builtin
feature of libgpiod.

w4tsn committed:
Add threading to watch for gpio interrupts

The underlaying library does not yet support callback based interrupt
watches, hence providing this functionality via threading.

w4tsn committed:
Fix ambigous line direction

w4tsn committed:
Add documentation of the module

w4tsn committed:
Add chardev config option

The libgpiod module requires a chardev added with linux kernel 4.8.
The default chardev is /dev/gpiochip0. Subsequent devices follow this
schema: /dev/gpiochipN.

w4tsn committed:
Fix wrong schema update

w4tsn committed:
Remove bias from gpiod module

The bias setting (Pullup) was removed from the unstable API. It will be
reintroduced at a later point in time.

The bias setting was introduced upstream in linux kernel 5.5.

w4tsn committed:
Add configuration example for gpiod

flyte commented on pi_mqtt_gpio/__init__.py:
```diff
@@ -127,6 +127,10 @@
+      chip:
+        type: string
+        required: no
+        default: "/dev/gpiochip0"
```
This module-specific configuration should only exist within the module's `CONFIG_SCHEMA` constant.

w4tsn replied:
I've moved the schema definition into the module.

flyte commented on pi_mqtt_gpio/modules/__init__.py:
```diff
@@ -8,6 +8,7 @@
     "name": {"required": True, "empty": False},
     "module": {"required": True, "empty": False},
     "cleanup": {"required": False, "type": "boolean", "default": True},
+    "chip": {"required": False, "type": "string", "default": "/dev/gpiochip0"},
```
As above, this should be in the module's `CONFIG_SCHEMA` constant instead.

w4tsn replied:
I've moved the schema definition into the module.

flyte commented on pi_mqtt_gpio/modules/gpiod.py:
```diff
@@ -0,0 +1,99 @@
+        config.consumer = 'pi-mqtt-gpio'
+        config.request_type = edge
+        
+        t = Thread(target=self._event_detect, args=(pin,self.interrupt_callback,))
```
Might be worth assigning this thread to `self.interrupt_thread` and then triggering an [event](https://docs.python.org/3/library/threading.html#event-objects) to stop it in `cleanup()`.

w4tsn replied:
I've rewritten this part to use it's own thread class, still not doing proper cleanup() handling. I had to notice while going at this again, that I couldn't get the release of a line to work by code (exiting the program releases the line). It might be viable to remove this implementation for now and not support callbacks for the first iteration until I've figured out how it's properly done in libgpiod

flyte commented on requirements.txt:
```diff
@@ -2,4 +2,4 @@ paho-mqtt
 enum34
 cerberus
-
+gpiod
```
This should only be in the module's `REQUIREMENTS` tuple so we don't have to install all requirements for all modules when installing the main program.

w4tsn replied:
I've moved this out of global requirements.

flyte reviewed (changes requested):
Thanks for the module! There are a couple of minor changes to be made before it can be merged. Feel free to make the changes yourself, or I'll do them at some point when I get around to it.

w4tsn commented:
I hope I'll eventually get to implement the requested changes in the next weeks :sweat_smile: 

Until then I've found another useful thing for this topic: Installing the [RPI.GPIO2 library] instead of RPI.GPIO will provide the python environment with a compatibility layer for the raspbian based syntax / bindings to the upstream mainline bindings. This way one may use the same code as before but actually talk to the upstream driver.

While I think a native support for the direct libgpiod bindings is preferred this may actually be an easy way. We only have to specify some configuration / logic to install the alternative library when not running on raspbian. Any ideas @flyte ?

[RPI.GPIO2 library]: <https://pypi.org/project/RPi.GPIO2/>

john-tho commented:
@w4tsn Thank you very much for providing this module, and the RPi.GPIO2 hint.
It would be great to see this finished and merged.
I would really like to avoid using `/dev/mem` on RPi, and use this software on a wider range of devices (example: OpenWrt based router / access point)
Cheers

w4tsn committed:
move gpiod module config to CONFIG_SCHEMA constant

w4tsn committed:
remove gpiod from global requirements

w4tsn committed:
rewrite interrupt handler as thread class

This approach moves the interrupt wait logic into it's own thread.

This does not yet work as the previous solution did not too. The
previous solution didn't actually register the interrupt. The new
one does try to register the interrupt but fails, since the line is
already bound as an input type from the pin init method before the
interrupt is called. Tried approaches to free the pin before trying to
bind it again or to find out if the input should be an interrupt did
not work yet.

w4tsn committed:
Remove interrupt functionality

The interrupt code didn't work yet and needs some more refinement in
a second iteration.

w4tsn commented:
I've rebased the branch to latest develop, tackled the open questions and removed the interrupt code since it didn't work and wasn't easy to finish.

I don't know why the schema checks are failing because I removed my additions which should bring it back to upstream state.

This provides basic functionality for libgpiod access. I propose to merge it without the interrupt code and bring that in in a second MR, when I better understand how that is supposed to work from the upstream library.

w4tsn committed:
Revert "Remove interrupt functionality"

This reverts commit 6300cc54c6f9fd7f60c513d04cad41c46e8b7dfb.

w4tsn committed:
fix interrupt code and add functionality

This finishes the interrupt code and enables interrupt capabilities.

w4tsn committed:
refactor setup_pin function

Make it more readable / simplify.

w4tsn commented:
Just took the time to finish this up. I've verified on a RPI4 with Fedora IoT 33 that interrupts work with all three modes rising, falling and both.

What remains as open questionables:

* there is a hard-coded timeout of 10 secs in the listener code - the python libgpiod library expects a timeout and 10 secs seems sane
* no cleanups implemented - from my experience the locks on the lines get released as soon as the program is closed
* pullup settings are not supported, since they were only introduced in Linux 5.5 and their support in the libs is still unknown to me
* since setup_pin is called before setup_interrupt and it's not clear if the input is configured as interrupt when setup_pin is called, the line is requested in both, where it is first released in setup_interrupt - it works


From my perspective this is ready to merge.

w4tsn commented:
@john-tho fyi: I've setup a container image as `quay.io/repository/w4tsn/python-mqtt-gpio` which is built from https://git.shivering-isles.com/w4tsn/python-mqtt-gpio-container and the feature-branch from my personal fork https://git.shivering-isles.com/w4tsn/python-mqtt-gpio

flyte commented:
Excellent work! Sorry for my absence. I'll take a look at the travis checks now, but I don't think they're failing because of anything you've done.
I appreciate all of the explanation you've added to the PR :+1:
